Assets/Shopify/UIToolkit/Components/RemoteImageLoader.cs

```diff
@@ -0,0 +1,140 @@
+namespace Shopify.UIToolkit {
+    using UnityEngine;
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+
+    public interface ImageLoader {
+        string GetError();
+
+        Texture2D GetTexture();
+
+        Dictionary<string, string> GetResponseHeaders();
+
+        IEnumerator Load(string url, Dictionary<string, string> headers);
+    }
+
+    public class UnityImageLoader : ImageLoader {
+        private WWW _www;
+
+        public string GetError() {
+            return _www.error;
+        }
+
+        public Texture2D GetTexture() {
+            return _www.texture;
+        }
+
+        public Dictionary<string, string> GetResponseHeaders() {
+            return _www.responseHeaders;
+        }
+
+        public IEnumerator Load(string url, Dictionary<string, string> headers) {
+            _www = new WWW(url, null, headers);
+            yield return _www;
+        }
+    }
+
+    /// <summary>
+    /// A behaviour that fetches and caches a remote image (JPEG/PNG) from the web.!-- from the web.
+    /// </summary>
+    public class RemoteImageLoader : MonoBehaviour {
+
+        private class HTTPHeaderFields {
+            public const string LAST_MODIFIED = "Last-Modified";
+            public const string STATUS_CODE = "STATUS";
+            public const string IF_MODIFIED_SINCE = "If-Modified-Since";
+        }
+
+        /// <summary>
+        /// The resulting error after downloading a web image.
+        /// </summary>
+        /// <returns></returns>
+        public string error {
+            get {
+                return _error;
+            }
+        }
+
+        /// <summary>
+        /// The resulting Texture2D web image after downloading from the given URL.
+        /// </summary>
+        /// <returns></returns>
+        public Texture2D texture {
+            get {
+                return _texture;
+            }
+        }
+
+        private const int HTTP_STATUS_NOT_MODIFIED = 304;
+
+        private WebImageCache _imageCache;
+        private string _error;
+        private Texture2D _texture;
+        private ImageLoader _loader;
+
+        public void SetImageLoader(ImageLoader loader) {
+            _loader = loader;
+        }
+
+        /// <summary>
+        /// Downloads the given web image located at the URL.
+        /// </summary>
+        /// <param name="url">URL of the image resource to download and cache.</param>
+        /// <param name="cache">Determine if the downloaded resource should be cached or not.</param>
+        /// <returns>On completion, check texture/error properties for result.</returns>
+        public IEnumerator DownloadImageURL(string url, bool cache = true) {
```

Comment:
The implementor is responsible for enumerating this coroutine? Maybe we could make this a private method and create a public method with optional callbacks that fire after this coroutine runs? That way the primary use case is more "fire and forget" with an image URL...

Reply:
I wasn't sure what the most common way of interfacing this would be but callbacks work for me!

Reply:
yea my intuition is that a dev will probably want all of the network/async concern hidden and will pretty much 90% of the time just call `RemoteImageLoader.LoadImageURL(url)` and forget about it...